PowerEditor/src/Notepad_plus.cpp

```diff
@@ -2497,6 +2498,300 @@ void Notepad_plus::setUniModeText()
 	_statusBar.setText(uniModeTextString.c_str(), STATUSBAR_UNICODE_TYPE);
 }
 
+bool isUrlSchemeStartChar(TCHAR const c)
+{
+	return ((c >= 'A') && (c <= 'Z'))
+		|| ((c >= 'a') && (c <= 'z'));
+}
+
+bool isUrlSchemeDelimiter(TCHAR const c) // characters allowed immedeately before scheme
+{
+	return   ! (((c >= '0') && (c <= '9'))
+			 || ((c >= 'A') && (c <= 'Z'))
+			 || ((c >= 'a') && (c <= 'z'))
+			 ||  (c == '_'));
+}
+
+bool isUrlTextChar(TCHAR const c)
+{
+	if (c <= ' ') return false;
+	switch (c)
+	{
+		case '"':
+		case '#':
+		case '\'':
+		case '<':
+		case '>':
+		case '?':
+		case '\0x7f':
+			return false;
+	}
+	return true;
+}
+
+bool isUrlQueryDelimiter(TCHAR const c)
+{
+	switch(c)
+	{
+		case '&':
+		case '+':
+		case '=':
+		case ';':
+			return true;
+	}
+	return false;
+}
+
+bool isUrlSchemeSupported (INTERNET_SCHEME s)
+{
+	switch (s)
+	{
+		case INTERNET_SCHEME_FTP:
+		case INTERNET_SCHEME_HTTP:
+		case INTERNET_SCHEME_HTTPS:
+		case INTERNET_SCHEME_MAILTO:
+			return true;

```

Comment:
The current scheme allow all the forms of URI: `[A-Za-z]+://...`
We can see in the future if users really need it
However, maybe we can consider at least the schemes of **file**, **data**, **ldap**, **telnet** and **irc**?
https://en.wikipedia.org/wiki/Uniform_Resource_Identifier

Reply:
OK, just checked the header, it's quite limited... in this case, why not to add INTERNET_SCHEME_FILE ?
It'll be nice to support such url:
file:///C:/Users/xxx/Downloads/sxxxcilier-fr-202010-invoice00239637.pdf

Reply:
Unfortunately, including the `file:` scheme doesn't mean to extend `isUrlSchemeSupported` only. File names require a different parsing:
- They can contain spaces. And spaces in file names should certainly be allowed when the file name is enclosed in double quotes and otherwise not.
- They can contain only one colon (located after the drive letter).
- They may not contain some other characters, which are allowed in `http:`-URLs.

I included some code to support the `file:` scheme, which works for me. I made it removable, if don't like it.

Reply:
> They can contain spaces. And spaces in file names should certainly be allowed when the file name is enclosed in double quotes and otherwise not.

File-names containing spaces are quite common.
And currently `file:///C:/Test 1.txt` is not parsed correctly.

If the fix is not simple, I'm wondering if removing `INTERNET_SCHEME_FILE` shouldn't be better.

Reply:
@Yaron10,

file names containing spaces are supposed to be enclosed in "double quotes", like in command lines. There is no other way to detect the end, except guessing.

Reply:
@Uhf7,

A file containing the `INTERNET_SCHEME_FILE` would usually be `file:///C:/Test 1.txt` (NOT `"file:///C:/Test 1.txt"`).
So the parsing would not work properly in many cases.

IMO this can cause confusion and removing `INTERNET_SCHEME_FILE` would be better.

Thank you.

Reply:
@Yaron10,
the correct way would be `file:///C:/Test%201.txt`. I added the special handling for the file scheme with quotes to allow really spaces in file names, but it seems not to be too intuitively.

Some ways to go here:
- Remove file scheme completely.
- Remove special handling for file scheme, then
  - Spaces have always to be coded as `%20`
  - It is possible to construct invalid file names, at least, if this method is used to open local files
- Allow the percent-encoding in the special handling for file scheme, then the spaces would work in quotes and the percent-encoding would work too.

I'm open to everything here.

Reply:
Continued in https://github.com/notepad-plus-plus/notepad-plus-plus/issues/8975.